flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/FixedBufferPool.java

```diff
@@ -0,0 +1,280 @@
+					throw new IllegalStateException("Buffer pool is destroyed.");
+				}
+
+				if (numberOfRequestedMemorySegments < numberOfRequiredMemorySegments) {
+					final MemorySegment segment = networkBufferPool.requestMemorySegment();
+
+					if (segment != null) {
+						numberOfRequestedMemorySegments++;
+						availableMemorySegments.add(segment);
+
+						continue;
+					}
+				}
+
+				if (askToRecycle) {
+					owner.releaseMemory(1);
+				}
+
+				if (isBlocking) {
+					availableMemorySegments.wait(2000);
+				} else {
+					return null;
+				}
+			}
+
+			return new Buffer(availableMemorySegments.poll(), this);
+		}
+	}
+
+	@Override
+	public void recycle(MemorySegment segment) {
+		synchronized (availableMemorySegments) {
+			if (isDestroyed) {
+				returnMemorySegment(segment);
+			} else {
+				BufferPoolListener listener = registeredListeners.poll();
+				if (listener == null) {
+					availableMemorySegments.add(segment);
+					availableMemorySegments.notify();
+				} else {
+					try {
+						boolean needMoreBuffers = listener.notifyBufferAvailable(new Buffer(segment, this));
+						if (needMoreBuffers) {
+							registeredListeners.add(listener);
+						}
+					} catch (Throwable ignored) {
+						availableMemorySegments.add(segment);
+						availableMemorySegments.notify();
+					}
+				}
+			}
+		}
+	}
+
+	/**
+	 * Destroy is called after the produce or consume phase of a task finishes.
+	 */
+	@Override
+	public void lazyDestroy() {
+		synchronized (availableMemorySegments) {
+			if (!isDestroyed) {
+				MemorySegment segment;
+				while ((segment = availableMemorySegments.poll()) != null) {
+					returnMemorySegment(segment);
+				}
+
+				registeredListeners.clear();
+
+				isDestroyed = true;
+			}
+		}
+
+		networkBufferPool.destroyBufferPool(this);
+	}
+
+	@Override
+	public boolean addListener(EventListener<Buffer> listener) {
+		return false;
+	}
+
+	@Override
+	public boolean addBufferPoolListener(BufferPoolListener listener) {
+		synchronized (availableMemorySegments) {
+			if (!availableMemorySegments.isEmpty() || isDestroyed) {
+				return false;
+			}
+
+			registeredListeners.add(listener);
+			return true;
+		}
+	}
+
+	@Override
+	public void setNumBuffers(int numBuffers) throws IOException {
+		checkArgument(numBuffers == numberOfRequiredMemorySegments,
+			"Buffer pool needs a fixed size of %s buffers, but tried to set to %s",
+			numberOfRequiredMemorySegments, numBuffers);
+	}
+
+	@Override
+	public String toString() {
+		synchronized (availableMemorySegments) {
+			return String.format("[required: %d, requested: %d, available: %d, listeners: %d, destroyed: %s]",
+				numberOfRequiredMemorySegments, numberOfRequestedMemorySegments,
+				availableMemorySegments.size(), registeredListeners.size(), isDestroyed);
+		}
+	}
+
+	// ------------------------------------------------------------------------
+
+	private void returnMemorySegment(MemorySegment segment) {
+		numberOfRequestedMemorySegments--;
```

Comment:
The `assert Thread.holdsLock(availableMemorySegments);` of the original `LocalBufferPool` class should be retained here as a safe-guard.